src/Features/Core/Portable/EditAndContinue/CommittedSolution.cs

```diff
@@ -427,11 +436,21 @@ await TryGetMatchingSourceTextAsync(log, sourceText, sourceFilePath, currentDocu
         }
     }
 
-    public void CommitSolution(Solution solution)
+    public void CommitChanges(Solution solution, ImmutableArray<ProjectId> projectsToStale, ImmutableArray<ProjectId> projectsToUnstale)
     {
         lock (_guard)
         {
             _solution = solution;
+            _staleProjects.AddRange(projectsToStale);
+            _staleProjects.RemoveRange(projectsToUnstale);
+
+            foreach (var documentId in _documentState.Keys)
+            {
+                if (projectsToUnstale.Contains(documentId.ProjectId))
+                {
+                    _documentState.Remove(documentId);
```

Comment:
i found this code odd to read.  basically, we're enumerating _documentState, but we're mutating as we enumerate?  is this safe?  is the enumeration ensured to hit all the docs, even if we mutate in place?

if there is no test for removing multiple docs, can you please add, to validate this?

Reply:
ok.  this is a normal Dictionary.  I'm nearly 100% sure it is not safe to mutate as you iterate.  You can make a temp array of docs to remove, then do a pass to remove them.  def add test :)

Reply:
Good catch. Forgot to make a copy of the keys.

Reply:
`Remove(this IDictionary, Func<Key, Value, bool> predicate);` would be useful.

Reply:
Oh, interesting - this actually works on .NET 9, but throws on .NET Framework:

![image](https://github.com/user-attachments/assets/a2de19c2-f4bc-4fb5-a202-3be66b88bea0)
![image](https://github.com/user-attachments/assets/ce23122d-45d4-4ec3-8264-75ba3d5ddc08)

That's why my tests passed locally and are failing in Desktop CI leg :)

Reply:
oh lol.  yeah. :)